File: android/src/main/java/org/ligi/gobandroid_hd/ui/review/GoGamePlayerActivity.java
package org.ligi.gobandroid_hd.ui.review;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.MotionEvent;
import android.view.View;

import org.ligi.gobandroid_hd.App;
import org.ligi.gobandroid_hd.InteractionScope;
import org.ligi.gobandroid_hd.R;
import org.ligi.gobandroid_hd.logic.Cell;
import org.ligi.gobandroid_hd.logic.GoGame;
import org.ligi.gobandroid_hd.logic.GoMove;
import org.ligi.gobandroid_hd.ui.GoActivity;
import org.ligi.gobandroid_hd.ui.GobanDroidTVActivity;
import org.ligi.gobandroid_hd.ui.alerts.GameForwardAlert;
import org.ligi.gobandroid_hd.ui.fragments.CommentAndNowPlayingFragment;
import org.ligi.gobandroid_hd.ui.ingame_common.SwitchModeHelper;
import org.ligi.tracedroid.logging.Log;

public class GoGamePlayerActivity extends GoActivity {

    private boolean autoplay_active = true;

    // timings in ms
    private final static int PAUSE_FOR_LAST_MOVE = 30000;
    private final static int PAUSE_BETWEEN_MOVES = 2300;
    private final static int PAUSE_BETWEN_MOVES_EXTRA_PER_WORD = 200;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        getBoard().setOnKeyListener(this);
        getBoard().do_actpos_highlight = false;

    }

    @Override
    protected void onStart() {
        if (autoplay_active) {
            new Thread(new autoPlayRunnable()).start();
        }
        super.onStart();
    }

    private Runnable mTimerProgressRunnable = new Runnable() {

        @Override
        public void run() {
            //setSupportProgress((int) (Window.PROGRESS_START + progress_to_display * (Window.PROGRESS_END - Window.PROGRESS_START)));
        }

    };

    float progress_to_display = 0.5f;

    /**
     * time in ms
     *
     * @param time
     */
    private void sleepWithProgress(int time) {
        try {
            long start_time = System.currentTimeMillis();

            while (System.currentTimeMillis() < start_time + time) {
                Thread.sleep(100);
                progress_to_display = 1f - ((float) (System.currentTimeMillis() - start_time + 1) / time);
                runOnUiThread(mTimerProgressRunnable);
            }

        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    class autoPlayRunnable implements Runnable {

        // private GoGame game;

        @Override
        public void run() {
            // game=;
            Log.i("gobandroid", "automove start" + getGame().getActMove().getNextMoveVariations().size());
            while (autoplay_active && (getGame().getActMove().hasNextMove())) {
                Log.i("gobandroid", "automove move" + getGame().getActMove().hasNextMove());

                runOnUiThread(new Runnable() {

                    @Override
                    public void run() {
                        GoMove next_mve = getGame().getActMove().getnextMove(0);
                        getGame().jump(next_mve);
                    }
                });
                Log.i("gobandroid", "automove move" + getGame().getActMove().hasNextMove());
                sleepWithProgress(calcTime());

            }
            Log.i("gobandroid", "automove finish " + autoplay_active);
            try {
                Thread.sleep(PAUSE_FOR_LAST_MOVE);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            Log.i("gobandroid", "automove asleep");

            if (!App.getInteractionScope().is_in_noif_mode()) {
                Intent next_intent = new Intent(GoGamePlayerActivity.this, GobanDroidTVActivity.class);

                if (autoplay_active) {
                    GoGamePlayerActivity.this.startActivity(next_intent);
                    GoGamePlayerActivity.this.finish();
                }
            } else {
                GoGamePlayerActivity.this.setResult(RESULT_OK);
                GoGamePlayerActivity.this.finish();
            }
        }

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        this.getMenuInflater().inflate(R.menu.ingame_review, menu);

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    protected void onStop() {
        autoplay_active = false;
        super.onStop();
    }

    public Fragment getGameExtraFragment() {
        return new CommentAndNowPlayingFragment();
    }

    @Override
    public byte doMoveWithUIFeedback(Cell cell) {
        // we want the user not to be able to edit in review mode
        return GoGame.MOVE_VALID;
    }

    @Override
    public boolean onKey(View v, int keyCode, KeyEvent event) {

        if (event.getAction() == KeyEvent.ACTION_DOWN)
            switch (keyCode) {
                case KeyEvent.KEYCODE_MEDIA_PAUSE:
                    SwitchModeHelper.startGame(this, InteractionScope.MODE_REVIEW);
                    return true;

                case KeyEvent.KEYCODE_MEDIA_PREVIOUS:
                case KeyEvent.KEYCODE_DPAD_LEFT:
                    if (getGame().canUndo()) {
                        getGame().undo();
                    }

                    return true;

                case KeyEvent.KEYCODE_DPAD_RIGHT:
                case KeyEvent.KEYCODE_MEDIA_NEXT:
                    GameForwardAlert.showIfNeeded(this, getGame());
                    return true;

                case KeyEvent.KEYCODE_DPAD_UP:
                case KeyEvent.KEYCODE_DPAD_DOWN:
                    return false;

            }
        return super.onKey(v, keyCode, event);
    }

    @Override
    public boolean isAsk4QuitEnabled() {
        return false;
    }

    public int countWords(String sentence) {
        int words = 0;
        if (!App.getInteractionScope().is_in_noif_mode())
            for (int i = 0; i < sentence.length(); i++)
                if (sentence.charAt(i) == ' ')
                    words++;

        return words;
    }

    private int calcTime() {
        int res = PAUSE_BETWEEN_MOVES;
        if (getGame().getActMove().hasComment())
            res += PAUSE_BETWEN_MOVES_EXTRA_PER_WORD * countWords(getGame().getActMove().getComment());
        return res;
    }

    @Override
    public boolean onTouch(View v, MotionEvent event) {
        return true; // this is a player - we do not want interaction
    }

}

File: android/src/main/java/org/ligi/gobandroid_hd/ui/review/GoGamePlayerActivity.kt
package org.ligi.gobandroid_hd.ui.review

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.os.SystemClock
import android.view.*
import kotlinx.android.synthetic.main.game.*
import org.ligi.gobandroid_hd.InteractionScope
import org.ligi.gobandroid_hd.R
import org.ligi.gobandroid_hd.logic.Cell
import org.ligi.gobandroid_hd.logic.GoGame.MoveStatus.VALID
import org.ligi.gobandroid_hd.ui.GoActivity
import org.ligi.gobandroid_hd.ui.GobanDroidTVActivity
import org.ligi.gobandroid_hd.ui.alerts.GameForwardAlert
import org.ligi.gobandroid_hd.ui.fragments.CommentAndNowPlayingFragment
import org.ligi.gobandroid_hd.ui.ingame_common.SwitchModeHelper

class GoGamePlayerActivity : GoActivity() {

    private val PAUSE_FOR_LAST_MOVE = 30000
    private val PAUSE_BETWEEN_MOVES = 2300
    private val PAUSE_BETWEN_MOVES_EXTRA_PER_WORD = 200

    private var autoplay_active = true

    inner class AutoPlayRunnable : Runnable {

        override fun run() {
            while (autoplay_active && game.actMove.hasNextMove()) {

                runOnUiThread {
                    val next_mve = game.actMove.getnextMove(0)
                    game.jump(next_mve)
                }
                sleepWithProgress(calcTime())

            }
            SystemClock.sleep(PAUSE_FOR_LAST_MOVE.toLong())

            if (!interactionScope.is_in_noif_mode) {
                val next_intent = Intent(this@GoGamePlayerActivity, GobanDroidTVActivity::class.java)

                if (autoplay_active) {
                    startActivity(next_intent)
                    finish()
                }
            } else {
                setResult(Activity.RESULT_OK)
                finish()
            }
        }

    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        go_board.setOnKeyListener(this)
        go_board.do_actpos_highlight = false

    }

    override fun onStart() {
        if (autoplay_active) {
            Thread(AutoPlayRunnable()).start()
        }
        super.onStart()
    }

    private val mTimerProgressRunnable = Runnable { setSupportProgress((Window.PROGRESS_START + progress_to_display * (Window.PROGRESS_END - Window.PROGRESS_START)).toInt()); }

    var progress_to_display = 0.5f

    /**
     * time in ms

     * @param time
     */
    private fun sleepWithProgress(time: Int) {
        val start_time = System.currentTimeMillis()

        while (System.currentTimeMillis() < start_time + time) {
            SystemClock.sleep(100)
            progress_to_display = 1f - (System.currentTimeMillis() - start_time + 1).toFloat() / time
            runOnUiThread(mTimerProgressRunnable)
        }
    }


    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        this.menuInflater.inflate(R.menu.ingame_review, menu)

        return super.onCreateOptionsMenu(menu)
    }

    override fun onStop() {
        autoplay_active = false
        super.onStop()
    }

    override val gameExtraFragment = CommentAndNowPlayingFragment()

    public override fun doMoveWithUIFeedback(cell: Cell?)
            // we want the user not to be able to edit in review mode
            = VALID

    override fun onKey(v: View, keyCode: Int, event: KeyEvent): Boolean {

        if (event.action == KeyEvent.ACTION_DOWN)
            when (keyCode) {
                KeyEvent.KEYCODE_MEDIA_PAUSE -> {
                    SwitchModeHelper.startGame(this, InteractionScope.Mode.REVIEW)
                    return true
                }

                KeyEvent.KEYCODE_MEDIA_PREVIOUS, KeyEvent.KEYCODE_DPAD_LEFT -> {
                    if (game.canUndo()) {
                        game.undo()
                    }

                    return true
                }

                KeyEvent.KEYCODE_DPAD_RIGHT, KeyEvent.KEYCODE_MEDIA_NEXT -> {
                    GameForwardAlert.showIfNeeded(this, game)
                    return true
                }

                KeyEvent.KEYCODE_DPAD_UP, KeyEvent.KEYCODE_DPAD_DOWN -> return false
            }
        return super.onKey(v, keyCode, event)
    }

    override fun isAsk4QuitEnabled() = false

    fun countWords(sentence: String) = if (interactionScope.is_in_noif_mode) {
        0
    } else {
        sentence.split(" ").size
    }

    private fun calcTime() = PAUSE_BETWEEN_MOVES + PAUSE_BETWEN_MOVES_EXTRA_PER_WORD * countWords(game.actMove.comment)

    override fun onTouch(v: View, event: MotionEvent) = true // this is a player - we do not want interaction

}
